framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1210,71 +1217,137 @@ int pcm_drain(struct pcm *pcm)
+ * @param[in] pcm A PCM handle
+ * @param[out] areas Returned mmap channel areas
+ * @param[out] offset Returned mmap area offset in frames
+ * @param[out] frames Mmap area portion size in frames
+ * @returns On success, zero is returned. On failure, a negative number returned
+ * @ingroup libtinyalsa-pcm
+ */
 int pcm_mmap_begin(struct pcm *pcm, void **areas, unsigned int *offset, unsigned int *frames)
 {
 	if (pcm == NULL || *areas == NULL || offset == NULL || frames == NULL) {
 		return -EINVAL;
 	}
-	int nframes = pcm_bytes_to_frames(pcm, pcm->next_size);
 
-	/* If data is not available, return -1 */
-	if (nframes == 0) {
+	int nframes = 0;
+	/* The next buffer to be used is pointed by mmap_idx */
+	struct ap_buffer_s *apb = pcm->pBuffers[pcm->mmap_idx];
+
+	/* If buffer is in enqueued state, return 0 */
+	if (apb->flags & AUDIO_APB_OUTPUT_ENQUEUED) {
 		*frames = 0;
-		return -1;
+		return 0;
 	}
 
-	*areas = pcm->next_buf->samp;
-	*offset = pcm->next_offset;
+	*areas = apb->samp;
 
-	if (*frames < nframes) {
-		pcm->next_size -= pcm_frames_to_bytes(pcm, *frames);
-		pcm->next_offset += pcm_frames_to_bytes(pcm, *frames);
+	if (pcm->flags & PCM_IN) {
+		*offset = pcm_bytes_to_frames(pcm, apb->curbyte);
+		nframes = pcm_bytes_to_frames(pcm, apb->nbytes - apb->curbyte);
 	} else {
-		*frames = pcm_bytes_to_frames(pcm, pcm->next_size);
-		pcm->next_size = 0;
-		pcm->next_offset = 0;
+		*offset = 0;
+		nframes = pcm_bytes_to_frames(pcm, apb->nmaxbytes);
+	}
+
+	if (*frames > nframes) {
+		*frames = nframes;
 	}
 
 	return 0;
 }
 
+/** Application indicates the completion of access to mmap area
+ * @param[in] pcm A PCM handle
+ * @param[in] offset Area offset in frames. This must be same as the offset returned by pcm_mmap_begin
+ * @param[in] frames Mmap area portion size in frames that application wishes to commit
+ * @returns On success, zero is returned. On failure, a negative number returned
+ * @ingroup libtinyalsa-pcm
+ */
 int pcm_mmap_commit(struct pcm *pcm, unsigned int offset, unsigned int frames)
 {
-	/* not used */
-	(void)offset;
-
 	if (pcm == NULL) {
 		return -EINVAL;
 	}
+
 	struct audio_buf_desc_s bufdesc;
+	struct ap_buffer_s *apb = pcm->pBuffers[pcm->mmap_idx];
 
-	if (pcm->next_size == 0) {
+	if (pcm->flags & PCM_OUT) {
+		/* In case of playback, we will always enqueue the buffer when
+		application calls commit, even if the buffer is not completely filled */
+		bufdesc.numbytes = pcm_frames_to_bytes(pcm, frames);
+	} else {
+		/* In record case, we will enqueue the buffer only after the application
+		reads all data present in buffer */
+		if (pcm_frames_to_bytes(pcm, offset + frames) < apb->nbytes) {
+			/* Application has not completely read the buffer. So, we only
+			update the offset in the buffer. The application will get the same buffer
+			when it calls mmap_begin() function again */
+			apb->curbyte = pcm_frames_to_bytes(pcm, offset + frames);
+			return frames;
```

Comment:
This return value is not in accordance with the return values defined in doxygen comments.

Reply:
Fixed